[WEAVE] OTEL attribute mapping table and threads example (#1780)
wandb/docs: weave/guides/tracking/otel.mdx

pip install openai openinference-instrumentation-openai opentelemetry-exporter-otlp-proto-http
```

<Warning>

> **Performance Recommendation**: Always use `BatchSpanProcessor` instead of `SimpleSpanProcessor` when sending traces to Weave. `SimpleSpanProcessor` exports spans synchronously and can cause performance issues. `BatchSpanProcessor` batches spans asynchronously, which is more efficient and recommended for production use.

Reviewer: ```suggestion
**Performance Recommendation**: Always use `BatchSpanProcessor` instead of `SimpleSpanProcessor` when sending traces to Weave. `SimpleSpanProcessor` exports spans synchronously, potentially impacting the performance of other workloads. These examples illustrate `BatchSpanProcessor`, which is recommended in production because it batches spans asynchronously and efficiently.
```
Totally optional and maybe not yet perfect. I wanted to tighten up the recommendation a bit. I'm tempted to remove "efficiently" entirely, because who knows?

```

The span attribute prefixes `gen_ai` and `openinference` are used to determine which convention to use, if any, when interpreting the trace. If neither key is detected, then all span attributes are visible in the trace view. The full span is available in the side panel when you select a trace.

> ## Organizing OTEL traces into threads

Reviewer: ```suggestion
## Organize OTEL traces into threads
```
I noticed that this page uses Title Case for many section headings. Maybe suggests a later refactor project across the docs.

Author: oh yeah, totally. I should have caught that.

> You can organize your OpenTelemetry traces into [Weave threads](/weave/guides/tracking/threads) by adding specific span attributes. This allows you to group related operations (like multi-turn conversations or user sessions) and visualize them in Weave's thread UI.

Reviewer: ```suggestion
Add specific span attributes to organize your OpenTelemetry traces into [Weave threads](/weave/guides/tracking/threads), then use Weave's Thread UI to analyze related operations like multi-turn conversations or user sessions in Weave's thread UI.
```
Maybe link to the thread UI? Do the Weave docs say UI or App?

Add the following attributes to your OTEL spans to enable thread grouping:

- `wandb.thread_id`: Groups spans into a specific thread
- `wandb.is_turn`: Marks a span as a conversation turn (appears as a row in the thread view)

> The following code shows four examples of how to organize OTEL traces into Weave threads using `wandb.thread_id` (to group related operations) and `wandb.is_turn` (to mark high-level operations that appear as rows in the thread view). The examples perform the following:

Reviewer: ```suggestion
The following code shows several examples of organizing OTEL traces into Weave threads. They use `wandb.thread_id` to group related operations,  and  use `wandb.is_turn` to view high level operations as rows in the thread view. Each example performs the followingmark high-level operations that appear as rows in the thread view). 
```

> * **Example 1**: Traces a basic single-turn thread

Reviewer: Super optional but I think this would look better as a little table, then maybe you could get away without "Traces" in each one They need periods.

* **Example 2**: Traces a multi-turn conversation sharing one thread ID

> * **Example 3**: Traces deeply nested operations where only the outermost span is marked as a turn

Reviewer: ```suggestion
* **Example 3**: Traces deeply nested operations and marks only the outermost span as a turn.
```

* **Example 4**: Traces background operations that belong to a thread but aren't turns

> <Accordion title="Click to open the code example">

Reviewer: Would it be cute to accordion each example separately? Then you could link the above.

    # Run examples
    example_1_basic_thread_and_turn()
    example_2_multiple_turns()
    example_3_complex_nested_structure()
    example_4_non_turn_operations()

    print("\n======================================")
    print("All examples completed!")
    print("\nKey points demonstrated:")
    print("1. wandb.thread_id groups related spans into threads")
    print("2. wandb.is_turn=true marks top-level operations as turns")
    print("3. Nested spans within turns should have is_turn unset or false")
    print("4. Multiple turns can share the same thread_id")
    print("5. Non-turn operations can be part of a thread")

if __name__ == "__main__":
    main()
```
</Accordion>

After sending these traces, you can view them in the Weave UI under the **Threads** tab, where they'll be grouped by `thread_id` and each turn will appear as a separate row.

## Attribute Mappings

Weave automatically maps OpenTelemetry span attributes from various instrumentation frameworks to its internal data model. When multiple attribute names map to the same field, Weave applies them in priority order, allowing frameworks to coexist in the same traces.

> ### Supported Frameworks

Reviewer: I think if this and the next one really belong in this page, it might be good to have an H2 called Appendix or something to contain them. It might be better to eventually split them out to a reference page.
